synapse/storage/util/id_generators.py

```diff
@@ -710,10 +710,17 @@ def get_current_token_for_writer(self, instance_name: str) -> int:
         # persisted up to position. This stops Synapse from doing a full table
         # scan when a new writer announces itself over replication.
         with self._lock:
-            return self._return_factor * self._current_positions.get(
+            pos = self._current_positions.get(
                 instance_name, self._persisted_upto_position
             )
 
+            # We also ensure that we always return at least the
+            # `persisted_upto_position` for ourselves, so that when we notify
+            # other workers about our position we give them the max valid value
+            # here so that nothing waits for us to advance.
+            pos = max(pos, self._persisted_upto_position)
+            return self._return_factor * pos
```

Comment:
Aside: return factor made me worried that we were scaling up or down stream ids. But this is only a sign (+/- 1)

Reply:
If self._return_factor is negative, will this do the right thing? I.e. is it possible that self._persisted_upto_position is negative and we should be doing `min` somehow? (Probably not?)

Reply:
Also. I wrote some words about streams [here](https://matrix-org.github.io/synapse/latest/development/synapse_architecture/streams.html#current-stream-id) which tried to define a per-writer current stream ID semiformally. Does this change make those words wrong?

Reply:
> If self._return_factor is negative, will this do the right thing? I.e. is it possible that self._persisted_upto_position is negative and we should be doing `min` somehow? (Probably not?)

Yeah, it should do the right thing here. Internally everything is kept as positive, then we apply `return_factor` to all positions before returning.



> Also. I wrote some words about streams [here](https://matrix-org.github.io/synapse/latest/development/synapse_architecture/streams.html#current-stream-id) which tried to define a per-writer current stream ID semiformally. Does this change make those words wrong?

I think that is still correct. I'm wondering whether we actually need to have two concepts: "current token" and a "last successful write"  one.

Reply:
It might be helpful if you could spell out an example that needs the max() fix for me.

Reply:
We do sometimes send it over federation? Do you want to jump on a call.

I've also seen a little bit of a problem with the current approach around waiting for stream updates for HTTP replication calls, so I'm having another look. (The code works, just might cause us to wait for longer than necessary)